Sources/NIOHTTP2/HTTP2FrameParser.swift

```diff
@@ -0,0 +1,835 @@
+    /// copy it into another buffer, with the corresponding allocation overhead.
+    ///
+    /// - Parameters:
+    ///   - frame: The frame to encode.
+    ///   - buf: Destination buffer for the encoded frame.
+    /// - Returns: An array containing zero or more additional buffers to send, in
+    ///            order. These may contain data frames' payload bytes, encoded
+    ///            header fragments, etc.
+    /// - Throws: Errors returned from HPACK encoder.
+    mutating func encode(frame: HTTP2Frame, to buf: inout ByteBuffer) throws -> IOData? {
+        // note our starting point
+        let start = buf.writerIndex
+        
+//      +-----------------------------------------------+
+//      |                 Length (24)                   |
+//      +---------------+---------------+---------------+
+//      |   Type (8)    |   Flags (8)   |
+//      +-+-------------+---------------+-------------------------------+
+//      |R|                 Stream Identifier (31)                      |
+//      +=+=============================================================+
+//      |                   Frame Payload (0...)                      ...
+//      +---------------------------------------------------------------+
+        
+        // skip 24-bit length for now, we'll fill that in later
+        buf.moveWriterIndex(forwardBy: 3)
+        
+        // 8-bit type
+        buf.write(integer: frame.payload.code)
+        // 8-bit flags -- we don't use padding when encoding in NIO, though
+        buf.write(integer: frame.flags.subtracting(.padded).rawValue)
+        // 32-bit stream identifier -- ensuring the top bit is empty
+        buf.write(integer: frame.streamID.networkStreamID!)
+        
+        func writePayloadSize(_ size: Int) {
+            var bytes: (UInt8, UInt8, UInt8)
+            bytes.0 = UInt8((size & 0xff_00_00) >> 16)
+            bytes.1 = UInt8((size & 0x00_ff_00) >>  8)
+            bytes.2 = UInt8( size & 0x00_00_ff)
+            withUnsafeBytes(of: bytes) { ptr in
+                _ = buf.set(bytes: ptr, at: start)
+            }
+        }
+        
+        let payloadStart = buf.writerIndex
+        
+        // frame payload follows, which depends on the frame type itself
+        switch frame.payload {
+        case .data(let data):
+            writePayloadSize(data.readableBytes)
+            return data
+            
+        case .headers(let headers, let priority):
+            if let priority = priority {
+                var dependencyRaw = UInt32(priority.dependency.networkStreamID ?? 0)
+                if priority.exclusive {
+                    dependencyRaw |= 0x8000_0000
+                }
+                buf.write(integer: dependencyRaw)
+                buf.write(integer: priority.weight)
+            }
+            
+            // HPACK-encoded data. Find the stream and its associated encoder/decoder
+            try self.headerEncoder.beginEncoding(allocator: self.allocator)
```

Comment:
Hrm. This suggests a possible API change, which is that the encoder could encode *into* a provided buffer. That will let us re-use this buffer here when necessary, which might be quite helpful.